File: android/app/src/main/java/com/purchasely/shaker/ui/screen/detail/DetailViewModel.kt
package com.purchasely.shaker.ui.screen.detail

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.purchasely.shaker.domain.repository.CocktailRepository
import com.purchasely.shaker.domain.repository.FavoritesRepository
import com.purchasely.shaker.domain.repository.PremiumRepository
import com.purchasely.shaker.domain.model.Cocktail
import com.purchasely.shaker.domain.usecase.ToggleFavoriteUseCase
import com.purchasely.shaker.purchasely.FetchResult
import com.purchasely.shaker.purchasely.PresentationHandle
import com.purchasely.shaker.purchasely.PurchaselyWrapper
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class DetailViewModel(
    private val repository: CocktailRepository,
    private val premiumRepository: PremiumRepository,
    private val favoritesRepository: FavoritesRepository,
    private val purchaselyWrapper: PurchaselyWrapper,
    private val toggleFavoriteUseCase: ToggleFavoriteUseCase,
    private val cocktailId: String
) : ViewModel() {

    private val _cocktail = MutableStateFlow<Cocktail?>(null)
    val cocktail: StateFlow<Cocktail?> = _cocktail.asStateFlow()

    val isPremium: StateFlow<Boolean> = premiumRepository.isPremium

    val favoriteIds: StateFlow<Set<String>> = favoritesRepository.favoriteIds

    // Signal Screen to display recipe presentation
    private val _requestRecipePresentation = MutableSharedFlow<PresentationHandle>()
    val requestRecipePresentation: SharedFlow<PresentationHandle> = _requestRecipePresentation.asSharedFlow()

    // Signal Screen to display favorites presentation
    private val _requestFavoritesPresentation = MutableSharedFlow<PresentationHandle>()
    val requestFavoritesPresentation: SharedFlow<PresentationHandle> = _requestFavoritesPresentation.asSharedFlow()

    init {
        _cocktail.value = repository.getCocktail(cocktailId)
        trackCocktailViewed()
    }

    private fun trackCocktailViewed() {
        // PURCHASELY: Increment a numeric counter attribute each time the user views a cocktail detail
        // Useful for triggering presentations after N views or segmenting engaged users
        // Docs: https://docs.purchasely.com/advanced-features/user-attributes
        purchaselyWrapper.incrementUserAttribute("cocktails_viewed")
        // PURCHASELY: Track the spirit of the last-viewed cocktail for personalized presentation content
        // Docs: https://docs.purchasely.com/advanced-features/user-attributes
        _cocktail.value?.spirit?.let { spirit ->
            purchaselyWrapper.setUserAttribute("favorite_spirit", spirit)
        }
    }

    fun isFavorite(): Boolean = favoritesRepository.isFavorite(cocktailId)

    fun toggleFavorite() {
        toggleFavoriteUseCase(cocktailId)
    }

    fun showRecipePresentation() {
        viewModelScope.launch {
            val result = purchaselyWrapper.loadPresentation(
                placementId = "recipe_detail",
                contentId = _cocktail.value?.id
            )
            when (result) {
                is FetchResult.Success -> {
                    _requestRecipePresentation.emit(result.handle)
                }
                is FetchResult.Client -> {
                    Log.d("DetailViewModel", "[Shaker] CLIENT presentation received for recipe_detail placement — build custom UI here")
                }
                else -> {}
            }
        }
    }

    fun showFavoritesPresentation() {
        viewModelScope.launch {
            val result = purchaselyWrapper.loadPresentation(placementId = "favorites")
            when (result) {
                is FetchResult.Success -> {
                    _requestFavoritesPresentation.emit(result.handle)
                }
                is FetchResult.Client -> {
                    Log.d("DetailViewModel", "[Shaker] CLIENT presentation received for favorites placement — build custom UI here")
                }
                else -> {}
            }
        }
    }

    fun onPresentationDismissed() {
        premiumRepository.refreshPremiumStatus()
    }
}

File: ios/Shaker/Screens/Detail/DetailViewModel.swift
import Foundation
import UIKit

class DetailViewModel: ObservableObject {

    @Published var cocktail: Cocktail?
    @Published var recipeFetchResult: FetchResult?
    @Published var favoritesFetchResult: FetchResult?

    private let repository = CocktailRepository.shared
    private let wrapper = PurchaselyWrapper.shared

    init(cocktailId: String) {
        cocktail = repository.cocktail(byId: cocktailId)
        trackCocktailViewed()
    }

    private func trackCocktailViewed() {
        // PURCHASELY: Increment a counter attribute each time the user views a recipe detail
        // Enables targeting users based on engagement depth (e.g., "viewed 5+ recipes")
        // Docs: https://docs.purchasely.com/advanced-features/user-attributes
        wrapper.incrementUserAttribute(forKey: "cocktails_viewed")
        if let spirit = cocktail?.spirit {
            // PURCHASELY: Record the spirit of the last-viewed cocktail as a user attribute
            // Allows personalized paywall content based on the user's preferred spirit category
            // Docs: https://docs.purchasely.com/advanced-features/user-attributes
            wrapper.setUserAttribute(spirit, forKey: "favorite_spirit")
        }
    }

    // MARK: - Prefetch Presentations

    func prefetchRecipePresentation(contentId: String) {
        Task {
            // PURCHASELY: Prefetch the "recipe_detail" paywall with a contentId for context
            // The contentId lets the paywall content reference the specific cocktail being viewed
            // Docs: https://docs.purchasely.com/quick-start/sdk-implementation/display-placements
            recipeFetchResult = await wrapper.loadPresentation(
                placementId: "recipe_detail",
                contentId: contentId
            ) { result in
                switch result {
                case .purchased, .restored:
                    PremiumManager.shared.refreshPremiumStatus()
                case .cancelled:
                    print("[Shaker] Recipe detail paywall cancelled")
                }
            }
        }
    }

    func prefetchFavoritesPresentation() {
        Task {
            // PURCHASELY: Prefetch the "favorites" paywall when the user views a detail screen
            // Docs: https://docs.purchasely.com/quick-start/sdk-implementation/display-placements
            favoritesFetchResult = await wrapper.loadPresentation(
                placementId: "favorites"
            ) { result in
                switch result {
                case .purchased, .restored:
                    PremiumManager.shared.refreshPremiumStatus()
                case .cancelled:
                    print("[Shaker] Favourites paywall cancelled")
                }
            }
        }
    }

    // MARK: - Display Paywalls

    func displayRecipePaywall(from viewController: UIViewController?) {
        guard case .success(let presentation) = recipeFetchResult else { return }
        wrapper.display(presentation: presentation, from: viewController)
    }

    func displayFavoritesPaywall(from viewController: UIViewController?) {
        guard case .success(let presentation) = favoritesFetchResult else { return }
        wrapper.display(presentation: presentation, from: viewController)
    }
}
